### PR TITLE
Fix a bug with an explicit typechecker import using hook

### PR DESCRIPTION
**Pull request by knyazer**
Hi, Patrick! 

I discovered an interesting problem, that is, while not being extremely critical, leads to meaningless error messages, and almost impossible to debug. While it could happen with any typechecker, it does not happen with typeguard, though happens with beartype. Here is a sample code:

```
import jaxtyping

%load_ext jaxtyping
%jaxtyping.typechecker beartype.beartype

#%% cell separation, so that hook can be applied

from beartype import *
def f(x: int):
    ...
f(1)
```
throws ```AttributeError: 'function' object has no attribute 'beartype'```

The reason for this is that when importing the beartype only the module is imported, so the annotation looks like ```@beartype.beartype```. The import of hook by beartype is done on top of the file/cell, thus any imports done inside file/cell override the original one, so the annotation becomes incorrect: from ```@beartype_module.beartype_func``` it becomes ```@beartype_func.beartype```. There are some related variants of the same problem, when, for example, beartype is applied as decorator, while being a module.

The solution is quite simple, just making sure that the name of the imported typechecker is different from anything non-adversarial. So, now instead of annotation with ```@beartype.beartype``` functions are annotated with ```@jaxtyping_beartype_beartype```.

### CODE REVIEW

**Comment by knyazer** [comment]
So the issue is that now there is no support for doing something like 
```
bearchecker = "beartype.beartype(conf=beartype.BeartypeConf(strategy=beartype.BeartypeStrategy.On))"
hook = install_import_hook("test.import_hook_tester_beartype_full", bearchecker)
``` 
because second argument is treated as a path, so it should not have any parentheses. Going to fix it, but this makes this PR much worse in terms of code quality.

**Comment by knyazer** [comment]
In fact, after some thought, it looks like the only solution would be to pass a function-definition to import hook, then do ```inspect.get_source``` and then after every ```import jaxtyped; import beartype ...``` redefine the passed function with some long and custom name, and then hope that this function was defined 'correctly', does not override anything important, etc. This approach seems simultaneously super-powerful (since it allows to define arbitrary decorators for anything jaxtyped), but also extremely fragile. (Note, that the current approach with passing a string seems even more fragile than inspecting source code of a function)

```
def function():
  import beartype
  return beartype.beartype

install_import_hook("...", function)
```
 and, in the file to which hook is applied:
```
import jaxtyping
def custom_jaxtyping_decorator():
    # stuff fetched via inspect.get_source
    import beartype
    return beartype.beartype

# user code...
```
For backwards compatibility the shorthand with passing a string should remain.


Writing this requires quite a bit of time, so before doing anything I wonder what do you, @patrick-kidger, think? Maybe there is a better approach that comes to your mind?

**Comment by patrick-kidger** [comment]
Eek! This is complicated. I see the name-conflict issue you're describing.

I think I've got a neater solution:
(a) we add a `decorator_lookup = {}` in `jaxtyping._import_hook`;
(b) we run
```python
def install_import_hook(..., typechecker: str):
    if typechecker not in decorator_lookup.keys():
        eval_string = f"import {typechecker.split('.', 1)[0]}\n" + typechecker
        decorator_lookup[typechecker] = eval(eval_string, ...)
        # The above should be thread-safe as long as `eval(eval_string)` is pure.
    ...
```
(c) we replace the existing AST-inserted import statements with `import jaxtyping`, and replace the AST-inserted decorator statements with `@jaxtyping._import_hook.decorator_lookup[typechecker]`.

And then, (d): also allow passing in a function definition instead, and simply place that in the `decorator_lookup` directly without the `eval`. FWIW the only reason that the import hook takes a string is because that string gets placed directly into the transformed AST -- but I've never really liked that approach as it's clearly quite fragile. We could consider adding a warning to that path to encourage folks to use the function-definition approach instead.

(FWIW I considered also something like generating `importlib.import_module("beartype").beartype` as the decorator directly, but I imagine the dynamic importing will slow things down on unjitted code.)

**Comment by knyazer** [comment]
I've implemented the solution with caching, there is only one questionable part, in my opinion: it is that the caching of custom-user-defined-functions typecheckers requires some sort of a persistent hash, and I decided to use a hash of ``inspect.getsource```. While it should work fine, it seems like relatively dumb approach. I am wondering if there is a neater solution that you could think of?

File: jaxtyping/_import_hook.py [hunk 1]
```diff
@@ -88,10 +90,58 @@ def _dot_lookup(*elements):
...
+        )
+        local_vars = {}
+        exec(string_to_eval, {}, local_vars)
+        _decorator_lookup[hash] = local_vars["f"]
+    elif callable(typechecker):
+        # If the typechecker is a function, treat it as a decorator
+        _decorator_lookup[hash] = typechecker
+    else:
+        # If it is None, ignore it silently (use dummy decorator)
+        # If it is anything else, we still want to ignore it.
```
Comment (patrick-kidger) [inline jaxtyping/_import_hook.py:141]: I think I'd suggest raising an error in this case.

File: jaxtyping/_import_hook.py [hunk 2]
```diff
@@ -385,6 +428,7 @@ def f(x: Float32[Array, "batch channels"]):
     else:
         raise RuntimeError("Cannot find a PathFinder in sys.meta_path")
 
+    _update_typechecker_decorator(typechecker)
```
Comment (patrick-kidger) [inline jaxtyping/_import_hook.py:431]: And add a warning to switch from the string-based approach to the function based approach?

File: test/test_import_hook.py [hunk 3]
```diff
@@ -65,6 +65,15 @@ def importhook_tempdir():
...
 
+def test_import_hook_with_function_typechecker(importhook_tempdir):
+    def f(x, *args, **kwargs):
+        import beartype
+
+        return beartype.beartype(x, *args, **kwargs)
+
+    _test_import_hook(importhook_tempdir, f)
+
+
```
Comment (patrick-kidger) [inline test/test_import_hook.py:76]: This test will fail if `beartype` isn't present (not all of the test environments I use have it installed); it needs a try-except block.

Also, I think probably best to duplicate all thes tests below, so that we also have
```python
_test_import_hook(importhook_tempdir, typeguard.typechecked)
_test_import_hook(importhook_tempdir, beartype.beartype)
_test_import_hook(importhook_tempdir, beartype.beartype(conf=beartype.BeartypeConf(strategy=beartype.BeartypeStrategy.On))
```
also explicitly tested.

**Comment by patrick-kidger** [comment]
So I've actually bumped into this issue before, and likewise haven't really found a better way. There are some edge-cases with working in the REPL, and potentially two functions in different modules that have the same source (but which produce different results due to different global or closure lookups). [This](https://github.com/patrick-kidger/equinox/blob/74a346b3173c1f6a3ac543867c0167c5c36ad03f/equinox/internal/_closure_to_pytree.py#L50) has been my solution.

File: jaxtyping/_import_hook.py (hunk 2 above)
Comment (knyazer) [reply jaxtyping/_import_hook.py:431]: Are you sure about this one? It seems quite hard to pass a function instead of a string, if you wanted to pass a string, so if somebody passes a function probably they know what they are doing. And then they would need to disable the warning, since it might annoy them..

I reckon that the function-style typechecker should be used only by advanced users, since in normal applications you never want to pass a function, so you point does make sense. But in any case, showing a warning if you do use such functionality seems a bit weird to me.

File: jaxtyping/_import_hook.py (hunk 1 above)
Comment (knyazer) [reply jaxtyping/_import_hook.py]: Yep, makes sense. Thanks!

File: test/test_import_hook.py (hunk 3 above)
Comment (knyazer) [reply test/test_import_hook.py:91]: Yeah, makes sense. Though maybe would be nice to also not require correct installation of typeguard: in my case I prefer to use exclusively beartype, since the newest version supports jaxtyping well. But that's the thing for another PR...

**Comment by knyazer** [comment]
I figured there is a lot of logic related to typechecker processing, so I just isolated everything into a class. Also now the hash for ```importlib``` and for addressing the decorators lookup is the same, which seems like a neat decision - now there are no problems with missing the cache, hopefully.

File: jaxtyping/_import_hook.py (hunk 2 above)
Comment (patrick-kidger) [reply jaxtyping/_import_hook.py]: Ah, so my thinking was that we'd make the function-passing form the recommended approach. I think it's a lot less magical to do so! Eval'ing code is pretty unpleasant business. The recommended approach becomes:
```python
import beartype
import jaxtyping
with jaxtyping.install_import_hook(..., beartype.beartype):
    ...
```

What's your thinking for the opposite?

Comment (knyazer) [reply jaxtyping/_import_hook.py]: Oh, no. I just completely missed the possibility to the thing you propose, because I thought that this just won't work.. But it does certainly work (unless users break beartype in the hooked file, but I think that is extremely unlikely).

**Comment by knyazer** [comment]
I reckon I fixed everything, the only thing left is to write the documentation, which I am going to leave to you. I am sure that your English skills are superior to mine :)

**Comment by knyazer** [comment]
So, @patrick-kidger, WDYT?

File: jaxtyping/_import_hook.py [hunk 4]
```diff
@@ -151,26 +205,24 @@ def visit_FunctionDef(self, node: ast.FunctionDef):
...
-    def __init__(self, *args, typechecker, **kwargs):
+    def __init__(self, *args, typechecker: _Typechecker, **kwargs):
         super().__init__(*args, **kwargs)
         self._typechecker = typechecker
-        self._typechecker_hash = hashlib.md5(
-            self._typechecker.encode("utf-8")
-        ).hexdigest()
 
     def source_to_code(self, data, path, *, _optimize=-1):
+        print("source to code.")
```
Comment (patrick-kidger) [inline jaxtyping/_import_hook.py:225]: I assume you meant to remove this!

File: jaxtyping/_import_hook.py [hunk 5]
```diff
@@ -88,14 +92,71 @@ def _dot_lookup(*elements):
...
+            exec(string_to_eval, {}, local_vars)
+            self.decorator = local_vars["f"]
+            self.hash = hash_string(typechecker)
+
+        elif callable(typechecker):
+            # If the typechecker is a function, treat it as a decorator
+            self.decorator = typechecker
+            try:
+                self.hash = hash_string(
+                    str(typechecker.__module__) + getsource(typechecker)
```
Comment (patrick-kidger) [inline jaxtyping/_import_hook.py:121]: I think we need the `__qualname__` here too, as `getsource` only includes the `__name__`.

Here's an example where two functions have the same `__module__`, `__name__`, and source, but different behaviour:
```python
import inspect

class A:
  x = 2

  def f(self):
    return self.x
  
class B:
  x = 3

  def f(self):
    return self.x

one = A().f
two = B().f
print(one.__module__, inspect.getsource(one))
print(two.__module__, inspect.getsource(two))
```

File: jaxtyping/_import_hook.py [hunk 6]
```diff
@@ -88,14 +92,71 @@ def _dot_lookup(*elements):
...
 
-def _str_lookup(string):
-    module = ast.parse(string)
-    (expr,) = module.body
-    return expr.value
+class _Typechecker:
+    lookup = {}
+
+    def __init__(self, typechecker):
+        self.typechecker = typechecker
```
Comment (patrick-kidger) [inline jaxtyping/_import_hook.py:99]: I don't think this is used anywhere, i.e. you can avoid saving it as an attribute.

File: jaxtyping/_import_hook.py [hunk 7]
```diff
@@ -88,14 +92,71 @@ def _dot_lookup(*elements):
...
+            # If the typechecker is a function, treat it as a decorator
+            self.decorator = typechecker
+            try:
+                self.hash = hash_string(
+                    str(typechecker.__module__) + getsource(typechecker)
+                )
+            except OSError:
+                # OSError means the function was defined in REPL
+                # We want to miss the importlib cache, so let's use the default hash
+                self.hash = str(typechecker.__hash__())
```
Comment (patrick-kidger) [inline jaxtyping/_import_hook.py:126]: just `str(hash(typechecker))`?

File: jaxtyping/_import_hook.py [hunk 8]
```diff
@@ -88,14 +92,71 @@ def _dot_lookup(*elements):
...
+        elif typechecker is None:
+            # If it is None, ignore it silently (use dummy decorator)
+            self.decorator = lambda x, *_, **__: x
```
Comment (patrick-kidger) [inline jaxtyping/_import_hook.py:129]: Likewise I don't think `self.decorator` is used anywhere, and doesn't need to be saved as an attribute.

File: jaxtyping/_ipython_extension.py [hunk 9]
```diff
@@ -17,7 +17,7 @@
 # IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 # CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 
-from ._import_hook import _JaxtypingTransformer
+from ._import_hook import _JaxtypingTransformer, _Typechecker
```
Comment (patrick-kidger) [inline jaxtyping/_ipython_extension.py:20]: Nit: I've realised that neither of these should have underscores -- the fact that we're importing them means they're clearly not private.

File: test/test_import_hook.py [hunk 10]
```diff
@@ -74,49 +74,133 @@ def _test_import_hook(importhook_tempdir, typechecker):
     assert counter + 1 == jaxtyping._test_import_hook_counter
 
 
-def test_import_hook_typeguard_old(importhook_tempdir):
-    _test_import_hook(importhook_tempdir, ("typeguard", "typechecked"))
+# Tests start here...
```
Comment (patrick-kidger) [inline test/test_import_hook.py:77]: ?

File: test/test_import_hook.py [hunk 11]
```diff
@@ -74,49 +74,133 @@ def _test_import_hook(importhook_tempdir, typechecker):
...
+    try:
+        import beartype  # noqa: F401,F811
+    except ImportError:
+        pytest.skip("Beartype not installed")
+    else:
+
+        def f(x, *args, **kwargs):
+            import beartype  # noqa: F401,F811
+
+            return beartype.beartype(x, *args, **kwargs)
```
Comment (patrick-kidger) [inline test/test_import_hook.py:90]: What's up with this wrapping? Just `_test_import_hook(importhook_tempdir, beartype.beartype)` should suffice I think?

Unrelatedly, I've realised the try/except can be cleaned up into a `beartype = pytest.importorskip("beartype")`.

So I think the whole test should be:
```python
def test_import_hook_function_beartype(importhook_tempdir):
    beartype = pytest.importorskip("beartype")
    _test_import_hook(importhook_tempdir, beartype.beartype)
```

**Review by patrick-kidger — COMMENTED** [review]
Mostly just minor comments. I like the approach you've taken overall! This is a nice improvement to the import hook.

File: jaxtyping/_import_hook.py (hunk 5 above)
Comment (patrick-kidger) [reply jaxtyping/_import_hook.py:121]: Actually, I've just realised that I think it's worse than this.

Suppose someone has:
```python
def somefunction(x):
    return x

def mydecorator(x):
    return somefunction(x)

with jaxtyping.install_import_hook("foo", mydecorator):
    ...
```
and then runs this, so that the `.pyc` files are generated.

Suppose they then change the contents of `somefunction`, and run it again.

In this case the old `.pyc` files will be hit, because `mydecorator` does not appear to have changed. That's a bug!

Even worse, I think that might be a bug with the current set-up as well. If e.g. `beartype.beartype` changes to some other definition, then again we don't pick up on that difference.

How does Python itself handle cache invalidation?

Comment (knyazer) [reply jaxtyping/_import_hook.py:121]: Oh, that's a nice bug. The only solution that comes to my mind (after some googling) is to do something with
```
import dis
def x(y):
     return y + 1
def y(z):
     return x(z + 2)

dis.dis(y)
```
which prints:
```
0 LOAD_GLOBAL              0 (x)
2 LOAD_FAST                0 (z)
4 LOAD_CONST               1 (2)
6 BINARY_ADD
8 CALL_FUNCTION            1
10 RETURN_VALUE
```
But this particular function is a relatively bad fit for our cause, and, to do something smarter, I will have to spend more time digging in the CPython sources.

And, yeah, I have found almost no information about cache handling in Python, so this will also require some time to figure out. But I think a better route is to make custom implementation of ```dis.dis```.

Comment (knyazer) [reply jaxtyping/_import_hook.py:121]: What I mean is that while ```dis.dis``` processes only the function body, and all calls are isolated, I think it should be totally possible to detect ```CALL_FUNCTION``` and then replace it with output of the recursive call of ```dis.dis```.

Comment (patrick-kidger) [reply jaxtyping/_import_hook.py:121]: Actually, I'm now realising that it maybe isn't as bad as I thought. The modified AST doesn't have a direct `@mydecorator` call. It has a `@jaxtyping._import_hook._Typechecker.lookup[some_hash]` call. We *can* afford to cache the `.pyc` file so long as the value stored in `lookup` is correctly invalidated.

I'd need to think a bit harder about what that entails. I think it means that the current implementation isn't buggy, at least, though. If the AST just has something like `@beartype.beartype` then it will still look up the new version given an old `.pyc` file.

Some interesting links I came across whilst researching this:
- [Optional bytecode cache](https://fedoraproject.org/wiki/Changes/Python_Optional_Bytecode_Cache)
- [PEP 3147](https://peps.python.org/pep-3147)
- [PEP 552](https://peps.python.org/pep-0552/)
- [`importlib.util.source_hash`](https://docs.python.org/3/library/importlib.html#importlib.util.source_hash)

Comment (patrick-kidger) [reply jaxtyping/_import_hook.py:121]: So *maybe* what that means is that not only is the current implementation fine, but we could weaken the hash key to just `(typechecker.__module__, typechecker.__qualname__)`? Since that's all that is needed to determine the hash used in the AST, which will then dynamically look up the typechecker at runtime?

For safety I could see us restricting things to global functions only (`isinstance(typechecker, types.FunctionType) and typechecker.__closure__ is None`). I think it would still go wrong given two locally-defined functions, that happen to have the same code but different closed-over variables. (Or maybe we just try to include the closed-over variables in the hash as well? I think that might be difficult, they might be unhashable values, e.g. dictionaries.)

File: jaxtyping/_import_hook.py (hunk 4 above)
Comment (knyazer) [reply jaxtyping/_import_hook.py]: Oops..

File: jaxtyping/_import_hook.py (hunk 6 above)
Comment (knyazer) [reply jaxtyping/_import_hook.py]: Yep! Thanks

File: jaxtyping/_import_hook.py (hunk 7 above)
Comment (knyazer) [reply jaxtyping/_import_hook.py]: Thanks!

File: jaxtyping/_ipython_extension.py (hunk 9 above)
Comment (knyazer) [reply jaxtyping/_ipython_extension.py]: They are still sort of module-wide private, but I agree that changing the names makes sense :)

File: test/test_import_hook.py (hunk 10 above)
Comment (knyazer) [reply test/test_import_hook.py]: Oops.. And the comment just helps to visually separate stuff that is not tests, and stuff that is tests.

File: test/test_import_hook.py (hunk 11 above)
Comment (knyazer) [reply test/test_import_hook.py]: Yep, I actually just made it a fixture (though it forces me to use not the standard module names). And so that ruff does not complain, I am forced to write a bit more verbosely. I prefer to do this:
```
def test(..., beartype_or_skip):
    _test_import_hook(..., beartype_or_skip.beartype)
```
over
```
def test(..., beartype_or_skip):
      _test_import_hook(..., beartype.beartype) # noqa: blahblah
```.

Though maybe you have a different opinion?

File: jaxtyping/_import_hook.py (hunk 8 above)
Comment (knyazer) [reply jaxtyping/_import_hook.py]: Indeed, thank you!

File: jaxtyping/_import_hook.py (hunk 5 above)
Comment (knyazer) [reply jaxtyping/_import_hook.py]: Hmm, but why cannot we use locally-defined functions? I think when we do, the expected behavior is defined a bit weirdly. If we typecheck the same module, it is just an awkward decision overall: I don't see why would somebody would do that. And, if we typecheck different modules with different local functions, yes, that is an issue, because we will have only one entry in the ```lookup```, but we can include the module we typecheck in the hash. Then there is no issue, because hashes are different, so the function used by the decorator will be the correct one.


In other words, if we assume that typechecking same module with two different typecheckers never happens, then including the typechecked module name in the hash should work fine, shouldn't it? Also, as you noted, we have the most recent version of the typechecker as the decorator, because it loaded dynamically, so maybe we won't have any issues at all as the result?

Comment (patrick-kidger) [reply jaxtyping/_import_hook.py]: I don't think we can make that assumption though. People *are* going to do all kinds of crazy things here.

Also, I'd be a bit antsy about including the module name in the hash. Hashes are not injective functions. When we're only hashing a fairly limited collection of things (strings like `"beartype.beartype"`) then I think it's fine; anything more large-scale and ever-so-maybe we get a collision error.

Now that we've seen how complicated this is, I'm thinking we should consider returning to only supporting strings, a la `install_import_hook(..., "beartype.beartype")`. And fix the original issue just by following (a)-(c) in [my original comment](https://github.com/google/jaxtyping/pull/116#issuecomment-1734130592).

File: test/test_import_hook.py (hunk 11 above)
Comment (patrick-kidger) [reply test/test_import_hook.py]: No strong feelings; this seems fine to me.

File: jaxtyping/_import_hook.py (hunk 5 above)
Comment (knyazer) [reply jaxtyping/_import_hook.py]: Yeah, including the typechecked module name in the hash increases the odds of collision quite a bit... And I absolutely agree with your position about increasing chances of collisions: the collision could lead to silently working not as expected, which is extremely bad.

But do you really think that going back to string processing is the best solution we can do? It limits the ability of user too much (I guess), and I can think of tons of ways to introduce bugs, e.g. typos, or, in general, passing any invalid typechecker. 

Limitations of the current approach are indeed annoying, but to me they don't look annoying enough to abandon the idea of allowing the functions. Though, maybe, if we assume that the users are crazy, and we want to be safe from spurious issues about 'why does my typechecker does not work', we should make this approach not the recommended one, but an "advanced" one.

Supporting only strings looks to me like a way to effectively force every user to use either beartype or typeguard: I am usually too scared to change something in a string that is going to be ```eval```-ed, and, overall, I believe that is one of the worst anti-patterns out there.

Comment (patrick-kidger) [reply jaxtyping/_import_hook.py]: Well, anything invalid should still throw an error. I don't think strings introduce any more risk of a silent bug than in any other case.

Conversely when using functions, I *am* concerned about accidentally failing to invalidate the cache in some edge case. I very much do not want to debug that!

Overall it seems like the wrong balance to have substantially more complicated jaxtyping code, just to make an unusual edge case slightly easier.

(If you really do need a custom typechecker, then it's easy enough to just define it in some other module, and refer to it by string: `install_import_hook(..., "my_module.my_typechecker")`.)

Comment (knyazer) [reply jaxtyping/_import_hook.py]: Well, I don't want to admit it, but I guess your position is as good as mine: I care more about being "easy to extend" and you care more about "not silently failing". And when thinking about it in this fashion, string-only approach looks significantly better...


Anyways, thanks for clarifying your opinion! I really appreciate that you spent a bit of your time to help me understand why the string-only approach is preferred.

Comment (patrick-kidger) [reply jaxtyping/_import_hook.py]: No worries! Thanks for being such a good sport. I agree that both positions could reasonably be taken. :)

**Comment by knyazer** [comment]
To sum up:
* User-defined typechecker decorator becomes ```jaxtyping.Typechecker.lookup[md5.hash(typechecker)]```. MD5 hash is used so that importlib and the lookup dictionary have the same hash, so that if there is a cache miss, it is guaranteed to happen for both importlib and lookup.
* ```test/test_import_hook``` now uses fixtures that either provide an import or skip the test. Even more concise than before.


Patrick, WDYT?

File: jaxtyping/_import_hook.py [hunk 12]
```diff
@@ -76,8 +76,11 @@ def _optimized_cache_from_source(typechecker_hash, /, path, debug_override=None)
     # Version 5: Added support for string-based `typechecker` argument.
     # Version 6: optimization tag now depends on `typechecker` argument, so that
     #    changing the typechecker will hit a different cache.
+    # Version 7: Now allowing to pass functions as a typechecker, using md5 hash of the
+    #    source code as the hash.
```
Comment (patrick-kidger) [inline jaxtyping/_import_hook.py:80]: I think this statement is no longer true. :)

File: test/conftest.py [hunk 13]
```diff
@@ -48,3 +48,13 @@ def _getkey():
...
 
     return _getkey
+
+
+@pytest.fixture(scope="module")
+def beartype_or_skip():
+    yield pytest.importorskip("beartype")
+
+
+@pytest.fixture(scope="module")
```
Comment (patrick-kidger) [inline test/conftest.py:58]: IIUC, making these module-level fixtures will result in a whole file getting skipped if these aren't present?

**Review by patrick-kidger — COMMENTED** [review]
I think this LGTM! My two comments aside I'm happy to merge this. :)

File: test/conftest.py (hunk 13 above)
Comment (knyazer) [reply test/conftest.py:58]: No, not exactly: if beartype or typeguard are not found, only the tests that use this fixture will be skipped. For example, if I uninstall beartype I get this report after doing ```pytest```:
```56 passed, 30 skipped, 1 warning in 5.40s```.
Or, when doing ```pytest test/test_import_hook```: ```4 passed, 3 skipped in 0.14s```.

So, I think it should work fine. At some point I would also like to change the codebase to support adequately either only typeguard or beartype, but this is not for this PR.

File: jaxtyping/_import_hook.py (hunk 12 above)
Comment (knyazer) [reply jaxtyping/_import_hook.py]: Yep, nice catch!

**Comment by knyazer** [comment]
I switched the IPython tests to use typeguard instead of beartype, so that with beartype uninstalled they don't fail.

**Comment by patrick-kidger** [comment]
Marvellous! In that case: LGTM, and merged.

Thank you for this one ; it ended up being quite thorny to handle correctly. I appreciate the effort you put into this. :)
